### PR TITLE
[backend version] oauth should not be usable

### PR DESCRIPTION
**Pull request by macejmic**

### CODE REVIEW

File: app/lib/backend_version.rb [hunk 1]
```diff
@@ -20,6 +20,10 @@ def oidc_usable?(_)
       false
     end
 
+    def oauth_usable?(service)
```
Comment (hallelujah) [inline app/lib/backend_version.rb]: We should put a rolling update here so SaaS users should be able to still do it in API

Comment (hallelujah) [reply app/lib/backend_version.rb]: /cc @mayorova do you know when all SaaS users are migrated?

Comment (mayorova) [reply app/lib/backend_version.rb]: @hallelujah no, and I don't expect it will be too soon.

File: app/lib/backend_version.rb [hunk 2]
```diff
@@ -20,6 +20,10 @@ def oidc_usable?(_)
...
+      service.provider.provider_can_use?(:oauth_api) || service.oidc?
```
Comment (hallelujah) [inline app/lib/backend_version.rb]: I think it should be `&&`

File: test/unit/backend_version_test.rb [hunk 3]
```diff
@@ -26,22 +26,36 @@ def test_visible_versions_oidc
...
     assert versions.is_a?(Array)
     assert versions.exclude?('oidc')
+
+    rolling_update(:oauth_api, enabled: true)
+    versions = BackendVersion.usable_versions(service: service)
+    assert versions.include?('oauth')
+
+    rolling_update(:oauth_api, enabled: false)
+    versions = BackendVersion.usable_versions(service: service)
+    refute versions.include?('oauth')
```
Comment (hallelujah) [inline test/unit/backend_version_test.rb]: What if `service.backend_version == :oidc` already?

File: app/lib/backend_version.rb (hunk 2 above)
Comment (macejmic) [reply app/lib/backend_version.rb]: not really. then user would be able to set `oauth` only through `oidc` option.

this way, saas users will be able to set simple `oauth`.

`oauth` will be usable for onprem only if `oidc` option has been picked.

Comment (hallelujah) [reply app/lib/backend_version.rb]: But we do not want oauth at all in the on-prem version

[the remark this answers is not in the document]
Comment (macejmic) [reply app/lib/backend_version.rb]: what's the difference? :)

[the remark this answers is not in the document]
Comment (Martouta) [reply app/lib/backend_version.rb]: aaaaah you are right! i didn't know about the existence of this:
https://github.com/3scale/porta/blob/3edba6a8a7ab9a620592d48634c5a10324e175f1/app/models/service.rb#L84

Comment (hallelujah) [reply app/lib/backend_version.rb]: image

Comment (hallelujah) [reply app/lib/backend_version.rb]: Refactoring this functionality is complex so will be done later.

**Comment by hallelujah** [comment]
So @mayorova be careful that when we merge this, only switching from OIDC to OAuth back and forth in SaaS will be possible.
@macejmic can you deploy to a staging environment and fix the failing tests please

File: app/lib/backend_version.rb [hunk 4]
```diff
@@ -20,8 +20,12 @@ def oidc_usable?(_)
       false
     end
 
+    def oauth_usable?(service)
+      service.provider&.provider_can_use?(:oauth_api) || service.oidc?
```
[the remark this answers is not in the document]
Comment (macejmic) [reply app/lib/backend_version.rb]: yep, it's required. but within test environment nothing works as the rest of the world.

**Pull request merged**